# Taller de Regresión Lineal Orientado al Negocio

Este taller se centra en **cómo** un negocio (por ejemplo, un e-commerce) puede **interpretar** y **aprovechar** los resultados de la Regresión Lineal para **tomar decisiones**.

## Contexto de Negocio
Imagina que eres el/la **gerente de marketing** de una tienda en línea. Tienes un conjunto de datos con el comportamiento de los clientes y la **cantidad anual** que han gastado (`Yearly Amount Spent`). Tu objetivo es:
1. **Identificar** qué factores (tiempo en la app, duración de la sesión, etc.) influyen más en el gasto anual.
2. **Predecir** el gasto anual futuro de los clientes.
3. Traducir estos hallazgos en **acciones concretas** que aumenten el gasto de los clientes.

## Datos Disponibles
Se te proporciona un dataset con las columnas:
- **Email**
- **Address**
- **Avatar**
- **Avg. Session Length** (Promedio de duración de la sesión)
- **Time on App** (Tiempo en la App)
- **Time on Website** (Tiempo en el Sitio Web)
- **Length of Membership** (Antigüedad de membresía, en años)
- **Yearly Amount Spent** (Cantidad gastada anualmente)

Tu variable objetivo para predecir es: **Yearly Amount Spent**.

## 1. Carga y Exploración de Datos

### Tarea 1.1: Cargar los datos
1. Crea un DataFrame (por ejemplo, `df`) leyendo el archivo CSV o introduciendo manualmente la tabla que se te proporcionó.
2. Muestra las primeras 5 filas (`df.head()`), la forma (`df.shape`) y los tipos de datos (`df.info()`).

**Pregunta de enfoque de negocio:**
- ¿Qué **insights** iniciales obtienes con solo ver las variables? ¿Cómo podría eso impactar tus **estrategias de marketing**?

In [ ]:
!wget https://github.com/javierherrera1996/lecture_analytics/raw/main/Ecommerce%20_Customers.csv

--2025-03-06 17:00:51--  https://github.com/javierherrera1996/lecture_analytics/raw/main/Ecommerce%20_Customers.csv
Resolving github.com (github.com)... 140.82.113.3
Connecting to github.com (github.com)|140.82.113.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/javierherrera1996/lecture_analytics/main/Ecommerce%20_Customers.csv [following]
--2025-03-06 17:00:52--  https://raw.githubusercontent.com/javierherrera1996/lecture_analytics/main/Ecommerce%20_Customers.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 87360 (85K) [text/plain]
Saving to: ‘Ecommerce _Customers.csv’

Ecommerce _Customer 100%[===================>]  85.31K  --.-KB/s    in 0.008s  

2025-03-06 17:00:52 (10.8 MB/s) - ‘Eco

In [ ]:
# EJEMPLO de código:
import pandas as pd

# Supongamos que has guardado tu dataset en un archivo 'data.csv'
df = pd.read_csv('Ecommerce _Customers.csv')
df.head()



Email  \
0      mstephenson@fernandez.com   
1              hduke@hotmail.com   
2               pallen@yahoo.com   
3        riverarebecca@gmail.com   
4  mstephens@davidson-herman.com   

                                             Address            Avatar  \
0       835 Frank Tunnel\nWrightmouth, MI 82180-9605            Violet   
1     4547 Archer Common\nDiazchester, CA 06566-8576         DarkGreen   
2  24645 Valerie Unions Suite 582\nCobbborough, D...            Bisque   
3   1414 David Throughway\nPort Jason, OH 22070-1220       SaddleBrown   
4  14023 Rodriguez Passage\nPort Jacobville, PR 3...  MediumAquaMarine   

   Avg. Session Length  Time on App  Time on Website  Length of Membership  \
0            34.497268    12.655651        39.577668              4.082621   
1            31.926272    11.109461        37.268959              2.664034   
2            33.000915    11.330278        37.110597              4.104543   
3            34.305557    13.717514        36.721283              3.120179   
4            33.330673    12.795189        37.536653              4.446308   

   Yearly Amount Spent  
0           587.951054  
1           392.204933  
2           487.547505  
3           581.852344  
4           599.406092

### Tarea 1.2: Análisis Exploratorio (EDA) Básico
1. Ejecuta `df.describe()` para revisar las estadísticas descriptivas de las variables numéricas.
2. Realiza uno o dos **boxplots** o **histogramas** para visualizar la distribución de:
   - **Avg. Session Length**
   - **Time on App**
   - **Time on Website**
   - **Length of Membership**
   - **Yearly Amount Spent**

**Preguntas de negocio:**
1. Si detectas que, por ejemplo, la gente con **mayor** `Length of Membership` gasta más, ¿qué **estrategias** implementarías para retener y "graduar" a los nuevos clientes a miembros más "antiguos"?

En realidad al observar los datos vemos que no hay como tal una relación directa u obligatoria entre gasto y tiempo, pero sí sería útil utilizar una estrategia de fidelización para seguir contando con esas ventas y quizá obtener una retroalimentación para saber cómo mejorar.

2. Si notas que el **Time on Website** es muy bajo vs. **Time on App**, ¿implica que debas mejorar la experiencia en el sitio web para que la gente pase más tiempo allí?

En realidad podemos observar que la gente pasa más tiempo en el sitio web que el la App, por lo que sería bueno ver cómo optimizar la aplicación para que la gente pueda usarla en cualquier dispositivo y momento, además de mejorar depronto su interfaz y hacerla más intuitiva o interactiva para que más gente la pueda usar.

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

In [ ]:
df.head(30)

Email  \
0         mstephenson@fernandez.com   
1                 hduke@hotmail.com   
2                  pallen@yahoo.com   
3           riverarebecca@gmail.com   
4     mstephens@davidson-herman.com   
5            alvareznancy@lucas.biz   
6             katherine20@yahoo.com   
7                awatkins@yahoo.com   
8       vchurch@walter-martinez.com   
9                  bonnie69@lin.biz   
10            andrew06@peterson.com   
11           ryanwerner@freeman.biz   
12                knelson@gmail.com   
13            wrightpeter@yahoo.com   
14            taylormason@gmail.com   
15              jstark@anderson.com   
16              wjennings@gmail.com   
17         rebecca45@hale-bauer.biz   
18          alejandro75@hotmail.com   
19           samuel46@love-west.net   
20                megan33@gmail.com   
21                agolden@yahoo.com   
22            vstafford@hotmail.com   
23  denise22@hernandez-townsend.com   
24           youngbarbara@yahoo.com   
25          william25@mcconnell.com   
26         ijones@schaefer-carr.net   
27            heatherhall@yahoo.com   
28          tinasmith@martinez.info   
29        chasejennifer@hotmail.com   

                                              Address            Avatar  \
0        835 Frank Tunnel\nWrightmouth, MI 82180-9605            Violet   
1      4547 Archer Common\nDiazchester, CA 06566-8576         DarkGreen   
2   24645 Valerie Unions Suite 582\nCobbborough, D...            Bisque   
3    1414 David Throughway\nPort Jason, OH 22070-1220       SaddleBrown   
4   14023 Rodriguez Passage\nPort Jacobville, PR 3...  MediumAquaMarine   
5   645 Martha Park Apt. 611\nJeffreychester, MN 6...       FloralWhite   
6   68388 Reyes Lights Suite 692\nJosephbury, WV 9...     DarkSlateBlue   
7               Unit 6538 Box 8980\nDPO AP 09026-4941              Aqua   
8              860 Lee Key\nWest Debra, SD 97450-0495            Salmon   
9               PSC 2734, Box 5255\nAPO AA 98456-7482             Brown   
10  26104 Alexander Groves\nAlexandriaport, WY 282...            Tomato   
11              Unit 2413 Box 0347\nDPO AA 07580-2652            Tomato   
12  6705 Miller Orchard Suite 186\nLake Shanestad,...         RoyalBlue   
13   05302 Dunlap Ferry\nNew Stephaniehaven, MP 42268            Bisque   
14  7773 Powell Springs Suite 190\nSamanthaland, N...          DarkBlue   
15  49558 Ramirez Road Suite 399\nPhillipstad, OH ...              Peru   
16        6362 Wilson Mountain\nJohnsonfurt, GA 15169        PowderBlue   
17               8982 Burton Row\nWilsonton, PW 88606         OliveDrab   
18  64475 Andre Club Apt. 795\nPort Dannytown, PW ...              Cyan   
19  544 Alexander Heights Suite 768\nNorth Johnvie...     LightSeaGreen   
20          84426 Julia Vista\nNorth Teresa, KY 50756         PeachPuff   
21              PSC 2490, Box 2120\nAPO AE 15445-2876             Black   
22                   PSC 5723, Box 8159\nAPO AA 74738             Olive   
23                   USNS Cardenas\nFPO AA 85439-9449            Silver   
24  019 Elliott Tunnel Suite 190\nNicholsbury, WV ...             Wheat   
25  9495 Mary Fall Apt. 777\nGlassport, ND 17957-5596              Teal   
26  657 Judith Crossroad\nHancockchester, VI 75658...             Linen   
27  8522 Regina Port Suite 782\nPort Kaitlin, TX 5...   MediumSlateBlue   
28      40000 Ann Port Suite 474\nYoungberg, MS 03651      LemonChiffon   
29  9507 Robert Prairie Apt. 601\nEast Crystalview...        DarkOrchid   

    Avg. Session Length  Time on App  Time on Website  Length of Membership  \
0             34.497268    12.655651        39.577668              4.082621   
1             31.926272    11.109461        37.268959              2.664034   
2             33.000915    11.330278        37.110597              4.104543   
3             34.305557    13.717514        36.721283              3.120179   
4             33.330673    12.795189        37.536653              4.446308   
5             33.871

In [ ]:
df.describe()

Avg. Session Length  Time on App  Time on Website  \
count           500.000000   500.000000       500.000000   
mean             33.053194    12.052488        37.060445   
std               0.992563     0.994216         1.010489   
min              29.532429     8.508152        33.913847   
25%              32.341822    11.388153        36.349257   
50%              33.082008    11.983231        37.069367   
75%              33.711985    12.753850        37.716432   
max              36.139662    15.126994        40.005182   

       Length of Membership  Yearly Amount Spent  
count            500.000000           500.000000  
mean               3.533462           499.314038  
std                0.999278            79.314782  
min                0.269901           256.670582  
25%                2.930450           445.038277  
50%                3.533975           498.887875  
75%                4.126502           549.313828  
max                6.922689           765.518462

## 2. Selección de Variables
En este dataset, algunas columnas como **Email**, **Address** o **Avatar** no son útiles directamente para el modelo de regresión.

### Tarea 2.1: Elegir features
- Usarás como variables predictoras (features) las siguientes columnas:
  1. **Avg. Session Length**
  2. **Time on App**
  3. **Time on Website**
  4. **Length of Membership**
- La columna objetivo (target) será **Yearly Amount Spent**.

**Pregunta de negocio**:
- ¿Por qué estas columnas tienen sentido para **explicar** el gasto anual? (Piensa en la lógica, "mientras más tiempo en app, más oportunidades de compra").

En realidad estas columnas son muy útiles, porque así nos podemos dar cuenta de deistintos factores y de cómo estos afectan el gasto de los consumidores en nuestra empresa, quizá en este caso no estén directamente relacionadas variables como el tiempo de membresía, ó el tiempo en la app con el gasto que hacen, pero eso nos da opción a buscar dentro de otras variables y ver en qué podemos mejorar para aumentar el gasto.

In [ ]:
x=df[["Avg. Session Length","Time on App","Time on Website","Length of Membership","Yearly Amount Spent"]]

In [ ]:
x

Avg. Session Length  Time on App  Time on Website  Length of Membership  \
0              34.497268    12.655651        39.577668              4.082621   
1              31.926272    11.109461        37.268959              2.664034   
2              33.000915    11.330278        37.110597              4.104543   
3              34.305557    13.717514        36.721283              3.120179   
4              33.330673    12.795189        37.536653              4.446308   
..                   ...          ...              ...                   ...   
495            33.237660    13.566160        36.417985              3.746573   
496            34.702529    11.695736        37.190268              3.576526   
497            32.646777    11.499409        38.332576              4.958264   
498            33.322501    12.391423        36.840086              2.336485   
499            33.715981    12.418808        35.771016              2.735160   

     Yearly Amount Spent  
0             587.951054  
1             392.204933  
2             487.547505  
3             581.852344  
4             599.406092  
..                   ...  
495           573.847438  
496           529.049004  
497           551.620145  
498           456.469510  
499           497.778642  

[500 rows x 5 columns]

## 3. Dividir el Dataset en Entrenamiento y Prueba
Para evaluar la capacidad de nuestro modelo de generalizar, dividiremos los datos en **train (80%)** y **test (20%)**.

**Pregunta de negocio**:
- ¿Cómo explicarías, en términos empresariales, que "apartamos" datos para prueba y no los usamos en el entrenamiento? ¿Qué importancia tiene para "predecir" clientes futuros?

El hecho de "apartar" cierta cantidad de datos es útil para enseñarle al modelo cómo hacer su trabajo y que él mismo obtenga retroalimentación y se dé cuenta cuando lo está haciendo bien. Esto nos ayudará a ver en un futuro como el tipo de comportamiento que esperamos de un buen cliente (ya sea alguiebn que lleve poco tiempo pero pase mucho en la app y por lo cual gasta más; o alguien que no necesita pasar tiempo en el sitio web porque ya lleva mucho tiempo como cliente y sabe lo que quiere, confia y gasta más).  

In [ ]:
X = df[['Avg. Session Length', 'Time on App', 'Time on Website', 'Length of Membership']]
y = df['Yearly Amount Spent']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(x,y,test_size=0.2,random_state=42)

print(f"Entrenamiento: {X_train.shape[0]} muestras")
print(f"Prueba: {X_test.shape[0]} muestras")

Entrenamiento: 400 muestras
Prueba: 100 muestras


## 4. Entrenar el Modelo de Regresión Lineal
1. **Crea** un objeto `LinearRegression()`.
2. **Entrena** el modelo con los datos de entrenamiento.
3. **Obtén** el intercepto y los coeficientes resultantes.

**Pregunta de negocio**:
- Si el coeficiente de `Time on App` es grande y positivo, ¿cómo podrías usar esta información para crear **campañas** o **mejoras** en la aplicación que incentiven al usuario a pasar más tiempo allí?

Sería lo mismo, ver cómo se puede mejorar la interfaz o la aplicación para que a la gente le guste pasar tiempo en ella o investigar qué más pueden hacer dentro de la misma, además se podrían usar algún tipo de recompensas por compras hechas en la app o registros que se hagan en la misma.

In [ ]:
model = LinearRegression()

In [ ]:
model.fit(X_train, y_train)


LinearRegression()

In [ ]:
intercept = model.intercept_
coef = model.coef_[0]

print(f"Intercept (beta_0): {intercept}")
print(f"Coeficiente (beta_1): {coef}")

Intercept (beta_0): -4.547473508864641e-13
Coeficiente (beta_1): 8.001648500289803e-15


## 5. Evaluación del Modelo
1. **Predice** en el conjunto de prueba.
2. Calcula las **métricas** MSE, MAE y R² para verificar la calidad de tu modelo. No profundizaremos en residuos, solo en la idea de si el modelo "le atina" al gasto anual.

**Preguntas de negocio**:
1. ¿Qué interpretas si **R²** es cercano a 0.80 o 0.90? ¿El modelo es “bueno” para tus objetivos?

En este caso, el R² es exactamente 1.0 tanto en entrenamiento como en prueba, lo que significa que el modelo explica el 100% de la variabilidad de los datos.
Esto es demasiado bueno para ser real y sugiere que el modelo probablemente está sobreajustado o que los datos de entrenamiento y prueba son idénticos.

2. Si el **MAE** es, por ejemplo, 15 USD, ¿es aceptable ese margen de error para tu negocio?

En este caso, el MAE es extremadamente bajo:

Entrenamiento: 4.10 × 10⁻¹⁴ (prácticamente cero)
Prueba: 4.03 × 10⁻¹⁴ (también prácticamente cero)
Estos valores confirman que el modelo ajustó perfectamente los datos. Sin embargo, en la vida real siempre hay un poco de error debido a la variabilidad de los datos. Valores tan cercanos a cero pueden indicar un problema como los mencionados antes: sobreajuste o uso incorrecto de variables.

In [ ]:
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error

In [ ]:
y_pred_train = model.predict(X_train)
y_pred_test = model.predict(X_test)

In [ ]:
mse_train = mean_squared_error(y_train, y_pred_train)
r2_train = r2_score(y_train, y_pred_train)
mae_train= mean_absolute_error(y_train, y_pred_train)

In [ ]:
mse_test = mean_squared_error(y_test, y_pred_test)
r2_test = r2_score(y_test, y_pred_test)
mae_test = mean_absolute_error(y_test, y_pred_test)

In [ ]:
print("Conjunto de Entrenamiento:")
print(f"MSE: {mse_train}")
print(f"R^2: {r2_train}")
print(f"MAE: {mae_train}")
print("\nConjunto de Prueba:")
print(f"MSE: {mse_test}")
print(f"R^2: {r2_test}")
print(f"MAE: {mae_test}")

Conjunto de Entrenamiento:
MSE: 3.691616600944665e-27
R^2: 1.0
MAE: 4.106937012693379e-14

Conjunto de Prueba:
MSE: 3.392732981174528e-27
R^2: 1.0
MAE: 4.035882739117369e-14


## 6. Interpretación y Aplicación Práctica
1. **Observa** los coeficientes resultantes: ¿qué variable es la más influyente? ¿Cómo impacta eso en tu estrategia de negocios?
2. **Propuestas** de acción:
   - Si `Time on App` es el que más incide, podrías **invertir** más en mejoras de la app.
   - Si `Length of Membership` está ligado a más gasto, podrías **fidelizar** a los nuevos clientes con promociones.

**Preguntas de negocio**:
1. ¿Cómo comunicarías estos resultados a un equipo directivo que no maneja detalles estadísticos?

En lugar de hablar de coeficientes, podríamos decir que los clientes que usan más la app o que llevan más tiempo con la empresa tienden a gastar más. Esto significa que, si queremos aumentar los ingresos, deberíamos enfocarnos en mejorar la experiencia en la app y en estrategias para fidelizar a los clientes nuevos. De esta forma, el equipo directivo entendería fácilmente qué acciones podríamos tomar para mejorar los resultados del negocio.

2. ¿Qué **acciones concretas** aplicarías para aumentar el gasto anual de los clientes, basándote en los coeficientes del modelo?

1) Invertir en mejorar la app con mejor UX, notificaciones personalizadas y ofertas dentro de la app.

2)Implementar programas de fidelización, descuentos progresivos y beneficios exclusivos para clientes con más tiempo.



## 7. Conclusión: Caso de Negocio Real
1. Redacta un breve **informe** o explicación final que responda:
   - ¿Qué **variables** explican mejor el gasto anual?
   - ¿Qué recomendaciones propones para aumentar el **Yearly Amount Spent**?

El modelo indica que las variables más influyentes en el gasto anual de los clientes son la antigüedad de membresía, el tiempo en la app y, en menor medida, el tiempo en el sitio web. Para aumentar el gasto, se recomienda fomentar la lealtad con programas de fidelización, optimizar la experiencia en la app y personalizar estrategias de venta según el comportamiento del cliente. Sin embargo, el modelo tiene limitaciones, como la ausencia de factores externos y la suposición de linealidad en la relación entre variables. Aun así, proporciona predicciones valiosas para mejorar la estrategia de negocio.



2. **Limitaciones**: Menciona cualquier aspecto que el modelo no cubra (datos no considerados, linealidad asumida, etc.).

No considera factores externos (competencia, economía, tendencias del mercado).

Con estos pasos, podrás llevar la **visión estadística** a un **plan de acción** en tu estrategia de negocio.

# Resumen de las Tareas
1. **EDA Rápido**: Cargar y describir el dataset.
2. **Selección de Variables**: Explicar por qué se usan las columnas numéricas y se ignoran Email/Address.
3. **Entrenamiento**: Ajustar el modelo de Regresión con `Avg. Session Length`, `Time on App`, `Time on Website`, `Length of Membership`.
4. **Evaluación**: Observar MSE, MAE, R².
5. **Interpretación**: Explicar cómo usaría el negocio estos resultados.

## ¡Listo para Analizar!
Sigue estos pasos, responde las **preguntas de negocio** y, sobre todo, **conecta** tus hallazgos con **acciones** en el mundo real.

> **Nota**: Puedes consultar documentación o herramientas como ChatGPT si lo requieres, pero recuerda que las **respuestas** deben ser **tuyas**, pensando en el contexto de negocio y la lógica detrás de los coeficientes. ¡Éxito!